Engineering/Mobile/ios_bluetooth_low_energy/ios_bluetooth_low_energy.md

```diff
@@ -0,0 +1,376 @@
+
+## IN BLEProofCentral SOURCE
+
+* App request permission for Bluetooth event. for that, in info.plist, I added this :
+
+```
+    <key>NSBluetoothAlwaysUsageDescription</key>
+    <string>We use Bluetooth to show basic communication between Central and Peripheral</string>
+    <key>NSBluetoothPeripheralUsageDescription</key>
+    <string>We use Bluetooth to show basic communication between Central and Peripheral</string>
+    <key>UIApplicationSupportsIndirectInputEvents</key>
+```
+
+* Init BLE
+
+```
+    private func initBLE() {
+        // using DispatchQueue.main means we can update UI directly from delegate methods
+        bleCentral = CBCentralManager(delegate: self, queue: DispatchQueue.main)
+    }
+```
+
+* BLELifecycleState
+
+```
+    enum BLELifecycleState: String {
+    case bluetoothNotReady
+    case disconnected
+    case scanning
+    case connecting
+    case connectedDiscovering
+    case connected
+    }
+```
+
+When scan and connect the central to peripheral, implement count down 30s timeout.
+
+* Count down time
+
+```
+    private func startCheckError(duration: Int = 30) {
+        cancelTimer()
+        countDownNumber = duration
+        timer = Timer.scheduledTimer(timeInterval: 1,
+                                     target: self,
+                                     selector: #selector(updateTime),
+                                     userInfo: nil,
+                                     repeats: true)
+    }
+    
+    private func cancelTimer() {
+        if let _ = self.timer {
+            self.timer!.invalidate()
+            self.timer = nil
+        }
+    }
+    
+    @objc func updateTime() {
+        if countDownNumber <= 0 {
+            finishTimer()
+            return
+        }
+        countDownNumber -= 1
+    }
+    
+    private func finishTimer() {
+        cancelTimer()
+        print("stage ==== \(lifecycleState)")
+        if lifecycleState == .scanning {
+            showAlertRetry(title: "Notification", message: "Connect Fail, plz retry it.") {
+                self.switchConnect.isOn = false
+            } retryAction: {
+                self.bleScan()
+            }
+
+        }
+    }
+```
+
+* Handle data recevice from bluetooth device (peripheral)
+
+```
+    func peripheral(_ peripheral: CBPeripheral, didUpdateValueFor characteristic: CBCharacteristic, error: Error?) {
+        guard error == nil else {
+            appendLog("didUpdateValue error: \(String(describing: error))")
+            return
+        }
+        
+        let data = characteristic.value ?? Data()
+        let stringValue = String(data: data, encoding: .utf8) ?? ""
+        if characteristic.uuid == uuidCharForIndicate {
+            handleDataNotify(data)
+        }
+        appendLog("didUpdateValue '\(stringValue)'")
+    }
+```
+
+* Function write and read Characteristic
+
+```
+    func bleReadCharacteristic(uuid: CBUUID) {
```

Comment:
Same here, can you remove the leading space and specify the language for the block?